tools/compiler/src/processor.rs

```diff
@@ -60,21 +60,53 @@ pub async fn generate_registry() -> AppResult<()> {
     Ok(())
 }
 
+/// Given `ibc_data` describing a channel and `source_asset` on the source chain,
+/// compute the metadata for the asset when it is transported along the channel onto a Penumbra chain.
+fn transport_metadata_along_channel(ibc_data: &IbcConfig, source_asset: Metadata) -> Metadata {
+    // The `Metadata` structure doesn't allow modifying the internals, so drop to raw proto data
+    let mut pb_metadata: pb::Metadata = source_asset.into();
+    tracing::debug!(?pb_metadata, "original");
+
+    let prefix_channel = |x: &mut String| {
+        *x = format!("transfer/{}/{}", ibc_data.ibc_channel, x);
+    };
+
+    // Prefix the channel to the base denom.
+    prefix_channel(&mut pb_metadata.base);
+    // Prefix the channel to the display denom.
+    prefix_channel(&mut pb_metadata.display);
+    // Prefix the channel to all denom units.
+    for denom_unit in pb_metadata.denom_units.iter_mut() {
+        prefix_channel(&mut denom_unit.denom);
+    }
```

Comment:
We should apply the prefix across all denoms and units. Also, I think the Rust `Metadata` parsing will reject if the base and display don't index into `denom_units`.

Reply:
Ah, we seemed to miss this in our current registry too: https://github.com/penumbra-zone/web/blob/main/packages/constants/src/local-asset-registry.json#L176-L188. Good to know.